api/ombpdf/semdom.py

```diff
@@ -0,0 +1,186 @@
+
+    def _push_child(self, name, attrs=None, elinfo=None):
+        child = self.document.createElement(name)
+        if attrs is not None:
+            for key, value in attrs.items():
+                child.setAttribute(key, value)
+        if elinfo is not None:
+            self.elinfo[child] = elinfo
+        self.cursor.appendChild(child)
+        self.cursor = child
+        return child
+
+    def _pop_child(self, name=None):
+        if name is not None:
+            assert self.cursor.nodeName == name
+        self.cursor = self.cursor.parentNode
+
+    def _get_parent(self, name):
+        node = self.cursor.parentNode
+        while node.nodeName != name:
+            node = node.parentNode
+        return node
+
+    def _ensure_section(self):
+        if self.cursor == self.root:
+            self._push_child('sec')
+
+    def _add_text(self, text):
+        self.cursor.appendChild(self.document.createTextNode(text))
+
+    def _add_comment(self, text):
+        self.cursor.appendChild(self.document.createComment(text))
+
+    # Below are Writer methods.
+
+    def begin_document(self, doc):
+        self.document = minidom.parseString('<policy></policy>')
+        self.root = self.document.childNodes[0]
+        self.cursor = self.root
+
+    def end_document(self, doc):
+        pass
+
+    def begin_heading(self, heading):
+        while self.depth >= heading.level:
+            self._pop_child()
+        self._push_child('sec', {'title': ''})
+        self._push_child('heading')
+        self.muted = False
+
+    def end_heading(self, heading):
+        self._pop_child('heading')
+        self.muted = True
+
+    def begin_paragraph(self, p):
+        self.muted = False
+        self._ensure_section()
+        self._push_child('para')
+        self._push_child('content')
+
+    def end_paragraph(self, p):
+        self._pop_child('content')
+        self._pop_child('para')
+        self.muted = True
+
+    def begin_list(self, li):
+        self._ensure_section()
+        self._push_child('list', elinfo=ListInfo(self, li))
+
+    def end_list(self, li):
+        self._pop_child('list')
+
+    def begin_list_item(self, p):
+        self.elinfo[self.cursor].begin_list_item()
+        self.muted = False
+
+    def end_list_item(self, p):
+        self.elinfo[self.cursor].end_list_item()
+        self.muted = True
+
+    def create_footnote_citation(self, cit):
+        self._push_child('footnote_citation')
+        self._add_text(str(cit.number))
+        self.footnote_citations[cit.number] = self.cursor
+        self._pop_child('footnote_citation')
+
+    def begin_footnote_list(self, fl):
+        self.muted = True
+
+    def end_footnote_list(self, fl):
+        if self.footnote_citations:
+            unresolved = str(list(self.footnote_citations.keys()))
+            self._add_comment(
+                'Warning, unresolved footnote citations exist: '
+                f'{unresolved}'
+            )
```

Comment:
Should this be a logged warning rather than part of the XML?

Reply:
Hmm, at the time that I wrote this, the XML output was the main output I was paying attention to, so it made the most sense to leave it as a comment, but I'm not sure if a comment *or* a logging message is the best solution right now... For example, it might be best to add some kind of "validation" step later in the pipeline that checks for these sorts of things, because it's ultimately something we want to surface to the end-user who's importing a PDF.